pdns/remote_logger.cc

```diff
@@ -31,73 +87,70 @@ bool RemoteLogger::reconnect()
 #endif
     return false;
   }
-  d_connected = true;
   return true;
 }
 
-void RemoteLogger::busyReconnectLoop()
+void RemoteLogger::queueData(const std::string& data)
 {
-  while (!reconnect()) {
-    sleep(d_reconnectWaitTime);
+  if(!d_writer) {
+    d_drops++;
+    return;
+  }
+  std::unique_lock<std::mutex> lock(d_mutex);
+  if(d_writer) {
+    try {
+      d_writer->write(data);
+    }
+    catch(std::exception& e) {
+      //      cout << "Got exception writing: "<<e.what()<<endl;
+      d_drops++;
+      d_writer.reset();
+      close(d_socket);
+      d_socket = -1;
+    }
   }
 }
 
-void RemoteLogger::worker()
+
+void RemoteLogger::maintenanceThread()
+try
 {
 #ifdef WE_ARE_RECURSOR
   string threadName = "pdns-r/remLog";
 #else
   string threadName = "dnsdist/remLog";
 #endif
   setThreadName(threadName);
-  while(true) {
-    std::string data;
-    {
-      std::unique_lock<std::mutex> lock(d_writeMutex);
-      d_queueCond.wait(lock, [this]{return (!d_writeQueue.empty()) || d_exiting;});
-      if (d_exiting) {
-        return;
-      }
-      data = d_writeQueue.front();
-      d_writeQueue.pop();
-    }
-
-    if (!d_connected) {
-      busyReconnectLoop();
-    }
 
-    try {
-      uint16_t len = static_cast<uint16_t>(data.length());
-      sendSizeAndMsgWithTimeout(d_socket, len, data.c_str(), static_cast<int>(d_timeout), nullptr, nullptr, 0, 0, 0);
+  for(;;) {
+    if(d_exiting)
+      break;
+    
+    if(d_writer) {
+      std::unique_lock<std::mutex> lock(d_mutex);
+      if(d_writer) { // check if it is still set
+        //        cout<<"Flush"<<endl;
+        try {
+          d_writer->flush();
+        }
+        catch(std::exception& e) {
+          //          cout<<"Flush failed!"<<endl;
+          d_writer.reset();
+          close(d_socket);
+          d_socket = -1;
+        }
+      }
     }
-    catch(const std::runtime_error& e) {
-#ifdef WE_ARE_RECURSOR
-      g_log<<Logger::Info<<"Error sending data to remote logger "<<d_remote.toStringWithPort()<<": "<< e.what()<<endl;
-#else
-      vinfolog("Error sending data to remote logger (%s): %s", d_remote.toStringWithPort(), e.what());
-#endif
-      busyReconnectLoop();
+    else if(reconnect()) { // if it was zero, it will remain zero, we are the only ones setting it!
```

Comment:
Perhaps we could try to reconnect right away if we just closed the connection? It feels especially bad to wait one second since we are no buffering messages when we are not connected.